src/main.rs

```diff
@@ -0,0 +1,547 @@
+        }
+        let point = ppoprf::Point::from(input.as_slice());
+        let evaluation = state.server.eval(&point, epoch, prove)?;
+        points.push(BASE64.encode(evaluation.output.as_bytes()));
+    }
+    let response = RandomnessResponse { points, epoch };
+    debug!("send: {response:?}");
+    Ok(Json(response))
+}
+
+/// Process PPOPRF epoch and key requests
+async fn info(
+    State(state): State<OPRFState>,
+) -> Result<Json<InfoResponse>, Error> {
+    debug!("recv: info request");
+    let state = state.read()?;
+    let public_key = state.server.get_public_key().serialize_to_bincode()?;
+    let public_key = BASE64.encode(public_key);
+    let response = InfoResponse {
+        current_epoch: state.epoch,
+        next_epoch_time: state.next_epoch_time.clone(),
+        max_points: MAX_POINTS,
+        public_key,
+    };
+    debug!("send: {response:?}");
+    Ok(Json(response))
+}
+
+/// Advance to the next epoch on a timer
+async fn epoch_update_loop(state: OPRFState, config: &Config) {
+    let interval =
+        std::time::Duration::from_secs(config.epoch_seconds.into());
+    info!("rotating epoch every {} seconds", interval.as_secs());
+
+    let epochs = config.first_epoch..=config.last_epoch;
+    loop {
+        // Pre-calculate the next_epoch_time for the InfoResponse hander.
+        // This acquires a temporary write lock which should be dropped
+        // before sleeping.
+        {
+            let now = time::OffsetDateTime::now_utc();
+            let next_rotation = now + interval;
+            // Truncate to the nearest second.
+            let next_rotation = next_rotation
+                .replace_millisecond(0)
+                .expect("should be able to round to a fixed ms.");
+            let timestamp = next_rotation
+                .format(&Rfc3339)
+                .expect("well_known timestamp format should always succeed");
+            // Locking should not fail, but if it does we can't set the field
+            // back to None, so panic rather than report stale information.
+            let mut s = state
+                .write()
+                .expect("should be able to update next_epoch_time");
+            s.next_epoch_time = Some(timestamp);
+        }
+
+        // Wait until the current epoch ends.
+        tokio::time::sleep(interval).await;
+
+        // Acquire exclusive access to the oprf state.
+        // Panics if this fails, since processing requests with an
+        // expired epoch weakens user privacy.
+        let mut s = state.write().expect("Failed to lock OPRFState");
+
+        // Puncture the current epoch so it can no longer be used.
+        let old_epoch = s.epoch;
+        s.server
+            .puncture(old_epoch)
+            .expect("Failed to puncture current epoch");
+
+        // Advance to the next epoch.
+        let new_epoch = old_epoch + 1;
+        if epochs.contains(&new_epoch) {
+            // Server is already initialized for this one.
```

Comment:
It happens most of the time. There are two levels of key rotation: generation of a new `ppoprf::Server` covering a set of epochs, and then puncturing individual epoch tags when the epoch is passed. So after a new server is created, all the future epoch tags are already present, we just  need to change which one we accept.

Only when all the epochs have been punctured do we need to create a completely new set.

You're right about the `u8` overflow problem, though. Good catch! I guess I haven't let it roll-over completely since switching that.

Reply:
ah yes, that is obvious now. obviously I was reviewing this while tired :laughing: